### ISSUE TITLE
Latency spike in distributors but not in ingesters

### ISSUE THREAD

**Issue by jmichalek132** [grafana/mimir#9709]
#### Describe the bug

At random times we are observing latency spikes in distributors that last from 1-2 minutes to 1-2 hours. They came and go (resolve on their own).
Whey they happen we don't see any corresponding latency spike in ingesters.
We don't seem to be hitting cpu / memory limits in distributors and adding more distributor pods doesn't seem have any effect.

Mimir overview
![Image](https://github.com/user-attachments/assets/af5a44c3-e874-4640-a570-1de7086bf75d)

Mimir writes 

![Image](https://github.com/user-attachments/assets/f792499f-103c-42b2-9c1c-903c321ea513)

Mimir writes resources

![Image](https://github.com/user-attachments/assets/a6f38100-025a-4e65-a034-ac734d225406)

Mimir writes networking

![Image](https://github.com/user-attachments/assets/441bcdc6-8424-48d0-b9a0-128890e12b40)

(The gap in metrics is unrelated, our prometheus scraping mimir was stuck in terminating state)

From traces we can see the latency seems to be spent in distributor and not in ingesters.

![Image](https://github.com/user-attachments/assets/85fbf71d-da6f-4b7e-b647-eb3ccb3eebb3)

The time seems to be spent in between 2 span events:

![Image](https://github.com/user-attachments/assets/c287c0af-4623-4170-9e42-4cc8c29eec39)

We can see it affects requests from the prometheus replicas from HA pair which is not elected and the data from it is not written to ingesters.

![Image](https://github.com/user-attachments/assets/4a2f3cad-56c9-41f5-9d11-ffafa8435739)

In our dev env we deployed modified version of mimir with 2 extra span events trying to narrow down where is the time spent.
Source with the modifications https://github.com/grafana/mimir/pull/9707/files. 

![Image](https://github.com/user-attachments/assets/55d7c6df-1fd5-45e7-9060-8137cd5ff4df)

And the source of the latency seems to be https://github.com/grafana/mimir/pull/9707/files#diff-e290efac4355b20d1b6858649bd29946bab12964336247c96ad1370e51e4503bR248.

It happens to in both of our environments production and dev.
It seems to affected subset of our tenants. 

#### To Reproduce

These latency spike seem to happen at random, so not sure how to reproduce them.

1. Start Mimir 2.14
2. Perform Operations(Read/Write/Others)

#### Expected behavior

No unexpected latency spikes.

#### Environment

- Infrastructure: kubernetes v1.29.5 (AKS, azure)
- Deployment tool: helm chart mimir-distributed 5.5.0

#### Additional Context

We didn't find anything in the logs of mimir distributor, given it doesn't log each incoming request, and doesn't log much.

**Comment by aknuds1** [comment]
> And the source of the latency seems to be https://github.com/grafana/mimir/pull/9707/files#diff-e290efac4355b20d1b6858649bd29946bab12964336247c96ad1370e51e4503bR248.

As that line should be reading from the client connection, is it possible that the latency stems from the client? It certainly looks that way to me, especially considering the trace you included (~18 seconds spent reading the proto payload).

**Comment by aknuds1** [comment]
I suspect this should be a discussion and not an issue, since there's no indication of a Mimir bug yet.

**Comment by jmichalek132** [comment]
> I suspect this should be a discussion and not an issue, since there's no indication of a Mimir bug yet. The symptoms so far are of an operational issue.

Hi, sorry I forgot about discussions, do you have the permissions by any chances to change into one?

**Comment by jmichalek132** [comment]
> > And the source of the latency seems to be https://github.com/grafana/mimir/pull/9707/files#diff-e290efac4355b20d1b6858649bd29946bab12964336247c96ad1370e51e4503bR248.
> 
> As that line should be reading from the client connection, is it possible that the latency stems from the client? It certainly looks that way to me, especially considering the trace you included (~18 seconds spent reading the proto payload).

Yeah that's what I am also starting to suspect the interesting thing is we don't see any increased latency on the ingress (contour based on envoy) but this might be down to what exactly the envoy metrics measure.
